polkadot/node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -1562,10 +1543,85 @@ fn process_expired_collations(
 			"Collation expired",
 		);
 
+		// Report metrics for collations that were expired after being backed.
+		if let Some(latency) = expired_collation.backed() {
+			metrics.on_collation_backed(latency as f64);
+		}
 		metrics.on_collation_expired(age as f64, collation_state);
 	}
 }
 
+/// Process collations that may have been finalized on the relay chain.
+///
+/// Collations no more tracked after this call.
+async fn process_possibly_finalized_collations(
+	collations: Vec<CollationStats>,
+	last_finalized: (Hash, BlockNumber),
+	sender: &mut impl overseer::SubsystemSender<ChainApiMessage>,
+	metrics: &Metrics,
+) {
+	if collations.is_empty() {
+		return
+	}
+
+	let (last_hash, last_number) = last_finalized;
+	let mut blocks_to_request = collations
+		.iter()
+		.map(|stats| stats.relay_parent().1)
+		.filter(|n| *n != last_number) // No need to request `last_hash` for `last_number`
+		.collect::<Vec<_>>();
+	blocks_to_request.sort_unstable();
+	blocks_to_request.dedup();
+
+	gum::debug!(target: LOG_TARGET_STATS, ?blocks_to_request, "Collations possibly finalized on blocks");
+
+	let mut finalized = vec![(last_hash, last_number)];
+
+	for chunk in blocks_to_request.as_slice().chunks(MAX_PARALLEL_CHAIN_API_REQUESTS) {
+		let futures = chunk.iter().map(|&bn| {
+				let mut sender = sender.clone();
+				async move {
+					let (tx, rx) = oneshot::channel();
+					sender.send_message(ChainApiMessage::FinalizedBlockHash(bn, tx)).await;
+					match rx.await {
+						Ok(Ok(Some(bh))) => Some((bh, bn)),
+						_ => {
+							gum::warn!(target: LOG_TARGET_STATS, block_number = ?bn, "Can't request hash for the finalized block from ChainApi");
+							None
+						},
+					}
+				}
+			});
+		finalized.extend(futures::future::join_all(futures).await.into_iter().flatten());
+	}
+
+	for collation in collations {
+		if finalized.contains(&collation.relay_parent()) {
+			// Report metrics for finalized collations
+			if let Some(latency) = collation.backed() {
+				metrics.on_collation_backed(latency as f64);
+			}
+			if let Some(latency) = collation.included() {
+				metrics.on_collation_included(latency as f64);
+			}
+			gum::debug!(
+				target: crate::LOG_TARGET_STATS,
+				relay_parent = ?collation.relay_parent(),
+				head = ?collation.head(),
+				"Collation finalized, stop tracking",
+			);
+		} else {
+			// Omit collations built on forks: they're dropped but it's expected
+			gum::debug!(
```

Comment:
perhaps personal preference, but I would change to `continue`-early if `!finalized.contains(&collation.relay_parent())`.